### PR TITLE
DOCS: Improve documentation of the EVP_PKEY type

### PR DESCRIPTION
**Pull request by levitte**
This type was previously described in a note, which is hard to find
unless you already know where to look.

This change makes the description more prominent, and allows indexing
by adding it in the NAMES section.

The EVP_PKEY description is altered to conceptually allow an EVP_PKEY
to contain a private key without a corresponding public key.  This is
related to an OTC vote:

https://mta.openssl.org/pipermail/openssl-project/2020-December/002474.html

The description of EVP_PKEY for MAC purposes is amended to fit.

### CODE REVIEW

**Comment by levitte** [comment]
This PR is *only* about changing the documentation of EVP_PKEY itself.  Diverse implementation changes, which is also part of what was voted on, should happen in separate PR(s?).

File: doc/man3/EVP_PKEY_check.pod [hunk 1]
```diff
@@ -38,11 +38,6 @@ It is not necessary to call these functions after locally calling an approved ke
 generation method, but may be required for assurance purposes when receiving
 keys from a third party.
 
-In OpenSSL an EVP_PKEY structure containing a private key also contains the
-public key components and parameters (if any). An OpenSSL private key is
-equivalent to what some libraries call a "key pair". A private key can be used
-in functions which require the use of a public key or parameters.
-
```
Comment (beldmit) [inline doc/man3/EVP_PKEY_check.pod:45]: Does it make sense adding some text to HISTORY about this change?

Comment (levitte) [reply doc/man3/EVP_PKEY_check.pod:45]: Hmmm, perhaps?  It's not like any type or function has been *added* or *removed*, which is what we usually mention in the history section.  If you have an idea what to actually write in the history section, I'm all ears.

Comment (beldmit) [reply doc/man3/EVP_PKEY_check.pod:45]: I'd suggest mentioning that the presence of both public and private key for private keys was implied before 3.0.

Comment (levitte) [reply doc/man3/EVP_PKEY_check.pod:45]: Ah, good point

Comment (levitte) [reply doc/man3/EVP_PKEY_check.pod:45]: Done

File: doc/man3/EVP_PKEY_new.pod [hunk 2]
```diff
@@ -172,6 +199,10 @@ The EVP_PKEY_new_raw_private_key_ex(),
 EVP_PKEY_new_raw_public_key_ex() and
 EVP_PKEY_new_CMAC_key_ex() functions were added in OpenSSL 3.0.
 
+The documentation of B<EVP_PKEY> was amended in OpenSSL 3.0 to allow there to
+be the private part of the keypair without the public part, where this was
+previously implied to be disallowed.
```
Comment (t8m) [inline doc/man3/EVP_PKEY_new.pod:204]: Maybe add something like: `implied to be disallowed although the implication was not enforced in some cases.`

Comment (levitte) [reply doc/man3/EVP_PKEY_new.pod:204]: Not sure why that's necessary...  furthermore, I would say it wasn't enforced, period!  Backends have always been able to disregard that rule.

Comment (t8m) [reply doc/man3/EVP_PKEY_new.pod:204]: ok

**Comment by t8m** [comment]
doc-nits failure seems relevant

**Comment by levitte** [comment]
> doc-nits failure seems relevant

Fixed

**Comment by levitte** [comment]
> doc-nits failure seems relevant

I missed a detail.  NOW it should be fixed

**Comment by levitte** [comment]
CIs are happy

**Comment by levitte** [comment]
Ping?

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by levitte** [comment]
Merged 

68e91251826949430270edfbf6dd5f334eb4df57 DOCS: Improve documentation of the EVP_PKEY type

**Pull request closed without merging**